전체 스타일 변경 액션 추가

Eunsol0410 opened the pull request:
## 수정전
- 상태 변경이 잘 되는지 테스트가 필요합니다. 
- import JSON을 하기 위해서는 input으로 json만 받도록 하는 Material UI?를 찾아보면 좋을 것 같습니다.
- 맵 스타일 변경은 레이어를 전체 업데이트 하는 거라서.. 테스트 후 방법을 논의해봐야할 것 같습니다. 

## 수정후
- 테스트 후 에러가 있다는 걸 발견하고 해결했습니다! 😭
- Json input editor 같은 유사한 모듈을 계속 찾아봤지만, 타입스크립트 호환이 안되어서 그냥 textarea를 사용하고 JSON Validation해주는 함수를 구현하면 좋을 것 같습니다.

Eunsol0410 committed:
[feat] 스타일 전체 업데이트 액션 추가

- 기존 스토어 상태와 동일한 형태로 전달받도록 하였고, 대신 변경되는 값만 전달되도록 하였음

Eunsol0410 committed:
[feat] 스타일 전체 수정 reducer 구현

- 변경된 사항만 업데이트 하기 위해서 combine 함수 만들었음

Eunsol0410 committed:
[feat] 전체 스타일 변경 hook 구현

- input으로 전달된 값을 스토어로 dispatch 하도록 하였음.
- 아직 맵 스타일도 수정하는 부분이 빠져있음
- 전체 레이어 업데이트 시간이 걸릴 것 같아서.. 논의가 필요함

qkrdmstlr3 commented:
댓글먼저 남깁니다! 왠만하면 직접 만들어봐요!!!

qkrdmstlr3 commented on src/hooks/common/useWholeStyle.ts:
```diff
@@ -0,0 +1,26 @@
+import { RootState } from '../../store';
+import { setWholeStyle } from '../../store/style/action';
+import { WholeStyleActionPayload } from '../../store/common/type';
+import * as mapStyling from '../../utils/map-styling';
```
어디 사용되는 건지 궁금합니다!! 추측하면 추후에 색상을 변경하는 것일까요??(아까 말씀하셨던 일일히 색상을 변경하는 것과 비슷할 까요?)

Eunsol0410 replied:
import, 테마에서 사용할 수 있는 함수입니다. 
추후 색상 변경이 맵 스타일 변경을 말씀하시는 걸까요?!! 그런 거라면 맞습니다!

qkrdmstlr3 reviewed (approved):
고생하셨습니다! 먼저 리뷰 하나 남기고 좀있다 연속해서 남길게요!

qkrdmstlr3 commented on src/store/common/type.ts:
```diff
@@ -69,12 +69,40 @@ export interface ElementPropsType extends FeaturePropsType {
+
+export type WholeStyleActionPayload = {
+  [featureName in FeatureNameType]?: {
+    [subFeatureName: string]: ElemetnActionPayload;
```
오타!! Elemetn!

Eunsol0410 replied:
앗! 네!!!! 바로 수정하겠습니다

qkrdmstlr3 commented on src/store/style/manageCategories.ts:
```diff
@@ -0,0 +1,41 @@
+  const update = initialElementStyle;
+  if (element === 'labelIcon' && elementStyle[element]) {
+    (update[element] as StyleActionPayload) = {
+      ...update[element],
```
update라는 변수를 따로 둔 것은 린트가 화났기 때문인가요 아니면 명시적인 표현 때문인가요?

Eunsol0410 replied:
그건 lint에서 no-param-reassign이라는 에러가 발생해서 이렇게 했습니다.
파라미터로 전달된 값의 값을 바꾸면 그 에러가 발생합니다.

qkrdmstlr3 replied:
아... js에서 변수하나 줄일 수 있어서 좋아하는 방식이었는데 아쉽네요..

qkrdmstlr3 reviewed (approved):
지도를 어떻게 효율적으로 한번에 렌더링 할수 있을지 생각해보았는데, 저희 초기화 할 때도 거의 100개 정도 색상 지정해주긴 합니다... 동기를 비동기적으로 작업하게할수는 없을라나요 ㅋㅋ. 다행히도 로직 자체는 복잡하지 않아서 코드 깔끔하네요!!

qkrdmstlr3 commented:
주말동안에 테마를 개인별로 json형태로 만들어와서 클릭시에 적용하는 방식으로 가면 좋을 것 같습니다!

Eunsol0410 commented:
@qkrdmstlr3 동기를 비동기적으로 처리한다는게 어떤 의미일까요?! 사실 지금 구조에서는 모든 레이어의 5가지 요소를 바꾸니깐 3X레이어 개수만큼 실행하게 되겠죠?!

Eunsol0410 committed:
[chore] 오타 수정

qkrdmstlr3 commented:
그냥 그런 상상을 해봤습니다. 모든 색상 처리가 준비 시작 하면 여러개가 동시에 실행되는..? 멀티 스레드라고 해야되나.. 근데 js는 그게 안되니깐...

Eunsol0410 commented:
@qkrdmstlr3 근데 신재경 멘토님이 쓰레드 이야기를 몇번 하셨어서.. 저도 그게 계속 궁금했어요!!! 내일 질문리스트에 포함하면 좋을 것 같아요

Eunsol0410 committed:
[fix] 기존 함수 오류 수정

- 상태 객체의 subFeature와 element가 혼동되어서 함수를 잘못 구현했음

- 기존 combine 함수를 2개로 분리해서 element(section, labelText, labelIcon) 별로 처리하고 subElement가 있는 경우를 분리된 함수에서 처리하도록 하였음

Eunsol0410 committed:
[feat] import 모달창에 useWholeStyle Hook 추가

- textarea 값이 json 파싱 문제가 있다고 생각했는데, 그게 아니라 json 형식 자체의 문제였음
- 따라서 전달된 json의 형식이 알맞게 왔는지 validation하는 함수가 필요해보임

gitgitWi commented on src/components/Sidebar/SidebarModal/ImportModal.tsx:
```diff
@@ -133,6 +136,14 @@ function ImportModal({
           onChange={onInputChange}
         />
         <ModalOKButton onClick={onClickOK}>지도 가져오기</ModalOKButton>
+        <button
+          type="button"
```
사소한 거긴 하지만 button 태그 안에서 `type="button"` 을 지정해줘야 하나요?

gitgitWi reviewed (commented):
고생하셨습니다!

qkrdmstlr3 committed:
Merge branch 'feature/map-style' into feat/style
